lib/src/util/condition_scope_visitor.dart

```diff
@@ -0,0 +1,256 @@
+    _addElementToEnvironment(new _ConditionExpression(node.condition));
+    node.updaters.accept(this);
+    node.body?.accept(this);
+    _removeLocalEnvironment();
+    /* TODO: (casanueva) Uncomment this when check no breaks inside
+    _addElementToEnvironment(
+        new _ConditionExpression(node.condition, value: false));
+        */
+  }
+
+  @override
+  visitFunctionDeclaration(FunctionDeclaration node) {
+    _addLocalEnvironment();
+    _addElementToEnvironment(new _UndefinedAllExpression());
+    node.visitChildren(this);
+    _removeLocalEnvironment();
+  }
+
+  @override
+  visitIfStatement(IfStatement node) {
+    final elseUndefinedExpressions =
+        _visitElseStatement(node.elseStatement, node.condition);
+    _visitIfStatement(node);
+    elseUndefinedExpressions
+        .where((e) => e is _UndefinedExpression)
+        .forEach(_addElementToEnvironment);
+    _addExpressionIfStatementIsStopStatement(node.thenStatement,
+        new _ConditionExpression(node.condition, value: false));
+    _addExpressionIfStatementIsStopStatement(
+        node.elseStatement, new _ConditionExpression(node.condition));
+  }
+
+  @override
+  visitMethodDeclaration(MethodDeclaration node) {
+    _addLocalEnvironment();
+    _addElementToEnvironment(new _UndefinedAllExpression());
+    node.visitChildren(this);
+    _removeLocalEnvironment();
+  }
+
+  @override
+  visitPostfixExpression(PostfixExpression node) {
+    final operand = node.operand;
+    if (operand is SimpleIdentifier) {
+      _addElementToEnvironment(new _UndefinedExpression(operand.bestElement));
+    }
+    node.visitChildren(this);
+  }
+
+  @override
+  visitPrefixExpression(PrefixExpression node) {
+    final operand = node.operand;
+    if (operand is SimpleIdentifier) {
+      _addElementToEnvironment(new _UndefinedExpression(operand.bestElement));
+    }
+    node.visitChildren(this);
+  }
+
+  @override
+  visitVariableDeclaration(VariableDeclaration node) {
+    _addElementToEnvironment(new _UndefinedExpression(node.element));
+    node.visitChildren(this);
+  }
+
+  @override
+  visitWhileStatement(WhileStatement node) {
+    _addLocalEnvironment();
+    visitCondition(node.condition);
+    node.condition?.accept(this);
+    _addElementToEnvironment(new _ConditionExpression(node.condition));
+    node.body?.accept(this);
+    _removeLocalEnvironment();
+    /* TODO: (casanueva) Uncomment this when check no breaks inside
+    _addElementToEnvironment(
+        new _ConditionExpression(node.condition, value: false));
+        */
+  }
+
+  void _addElementToEnvironment(_ExpressionBox e) {
+    if (e != null && environments.isNotEmpty) {
+      environments.first.addFirst(e);
+    }
+  }
+
+  _addExpressionIfStatementIsStopStatement(
+      Statement statement, _ExpressionBox expressionBox) {
+    if (_isReturnStatement(statement) ||
+        _isBreakStatement(statement) ||
+        _isContinueStatement(statement)) {
+      _addElementToEnvironment(expressionBox);
```

Comment:
We do something similar for dead code detection. Check out https://github.com/dart-lang/sdk/blob/master/pkg/analyzer/lib/src/generated/resolver.dart#L2201. Our checking is a bit more complete, and you might be able to make use of it here (although we'd need to make it available).